File: app/src/main/java/fr/free/nrw/commons/upload/categories/CategoriesContract.java
package fr.free.nrw.commons.upload.categories;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import fr.free.nrw.commons.BasePresenter;
import fr.free.nrw.commons.Media;
import fr.free.nrw.commons.category.CategoryItem;
import java.util.List;

/**
 * The contract with with UploadCategoriesFragment and its presenter would talk to each other
 */
public interface CategoriesContract {

    interface View {

        void showProgress(boolean shouldShow);

        void showError(String error);

        void showError(int stringResourceId);

        void setCategories(List<CategoryItem> categories);

        void goToNextScreen();

        void showNoCategorySelected();

        /**
         * Gets existing category names from media
         */
        List<String> getExistingCategories();

        /**
         * Returns required context
         */
        Context getFragmentContext();

        /**
         * Returns to previous fragment
         */
        void goBackToPreviousScreen();

        /**
         * Shows the progress dialog
         */
        void showProgressDialog();

        /**
         * Hides the progress dialog
         */
        void dismissProgressDialog();

        /**
         * Refreshes the categories
         */
        void refreshCategories();
    }

    interface UserActionListener extends BasePresenter<View> {

        void searchForCategories(String query);

        void verifyCategories();

        void onCategoryItemClicked(CategoryItem categoryItem);

        /**
         * Attaches view and media
         */
        void onAttachViewWithMedia(@NonNull CategoriesContract.View view, Media media);

        /**
         * Clears previous selections
         */
        void clearPreviousSelection();

        /**
         * Update the categories
         */
        void updateCategories(Media media, String wikiText);

        LiveData<List<CategoryItem>> getCategories();

        void selectCategories();

    }


}

File: app/src/main/java/fr/free/nrw/commons/upload/categories/CategoriesContract.kt
package fr.free.nrw.commons.upload.categories

import android.content.Context
import androidx.lifecycle.LiveData
import fr.free.nrw.commons.BasePresenter
import fr.free.nrw.commons.Media
import fr.free.nrw.commons.category.CategoryItem

/**
 * The contract with with UploadCategoriesFragment and its presenter would talk to each other
 */
interface CategoriesContract {
    interface View {
        fun showProgress(shouldShow: Boolean)

        fun showError(error: String?)

        fun showError(stringResourceId: Int)

        /**
         * Show a cancelable AlertDialog with a given message.
         */
        fun showErrorDialog(message: String)

        fun setCategories(categories: List<CategoryItem>?)

        fun goToNextScreen()

        fun showNoCategorySelected()

        /**
         * Gets existing category names from media
         */
        fun getExistingCategories(): List<String>?

        /**
         * Returns required context
         */
        fun getFragmentContext(): Context

        /**
         * Returns to previous fragment
         */
        fun goBackToPreviousScreen()

        /**
         * Shows the progress dialog
         */
        fun showProgressDialog()

        /**
         * Hides the progress dialog
         */
        fun dismissProgressDialog()

        /**
         * Refreshes the categories
         */
        fun refreshCategories()

        /**
         * Navigate the user to Login Activity
         */
        fun navigateToLoginScreen()
    }

    interface UserActionListener : BasePresenter<View> {
        fun searchForCategories(query: String)

        fun verifyCategories()

        fun onCategoryItemClicked(categoryItem: CategoryItem)

        /**
         * Attaches view and media
         */
        fun onAttachViewWithMedia(view: View, media: Media)

        /**
         * Clears previous selections
         */
        fun clearPreviousSelection()

        /**
         * Update the categories
         */
        fun updateCategories(media: Media, wikiText: String)

        fun getCategories(): LiveData<List<CategoryItem>>

        fun selectCategories()
    }
}
